include/video.h

```diff
@@ -40,6 +41,168 @@ enum class IntegerScalingMode {
+
+	// The intended pixel aspect ratio of the video mode. Note this is not
+	// simply calculated by stretching 'width x height' to a 4:3 aspect
+	// ratio rectangle; it can be literally anything.
+	Fraction pixel_aspect_ratio = {};
+
+	// - For graphics modes, the first graphics standard the mode was
+	//   introduced in, unless there is ambiguity, in which case the emulated
+	//   graphics adapter (e.g. in the case of PCjr and Tandy modes).
+	// - For text modes, the graphics adapter in use.
+	GraphicsStandard graphics_standard = {};
+
+	// Colour depth of the video mode. Note this is *not* the same as the
+	// storage bit-depth; e.g., some 24-bit true colour modes actually store
+	// pixels at 32-bits with the upper 8-bits unused.
+	ColorDepth color_depth = {};
+
+	// True if this is a double-scanned mode on VGA (e.g. 200-line CGA and
+	// EGA modes and most sub-400-line (S)VGA & VESA modes)
+	bool is_double_scanned_mode = false;
+
+	bool is_cga_or_ega_mode_with_vga_palette = false;
+
+	constexpr bool operator==(const VideoMode& that) const
+	{
+		return (bios_mode_number == that.bios_mode_number &&
+		        is_graphics_mode == that.is_graphics_mode &&
+		        is_custom_mode == that.is_custom_mode &&
+		        width == that.width && height == that.height &&
+		        pixel_aspect_ratio == that.pixel_aspect_ratio &&
+		        graphics_standard == that.graphics_standard &&
+		        color_depth == that.color_depth &&
+		        is_double_scanned_mode == that.is_double_scanned_mode);
+	}
+
+	constexpr bool operator!=(const VideoMode& that) const
+	{
+		return !operator==(that);
+	}
+};
+
+std::string to_string(const VideoMode& video_mode);
+
+
+enum class PixelFormat : uint8_t {
+	// Up to 256 colours, paletted;
+	// stored as packed uint8 data
+	Indexed8 = 8,
+
+	// 32K high colour, 5 bits per red/blue/green component;
+	// stored as packed uint16 data with highest bit unused
+	BGR555 = 15,
+	//
+	// 65K high colour, 5 bits for red/blue, 6 bit for green;
+	// stored as packed uint16 data
+	BGR565 = 16,
+	//
+	// 16.7M (24-bit) true colour, 8 bits per red/blue/green component;
+	// stored as packed 24-bit data
+	BGR888 = 24,
+	//
+	// 16.7M (32-bit) true colour; 8 bits per red/blue/green component;
+	// stored as packed uint32 data with lowest 8 bits unused
+	BGRX8888 = 32
+};
+
+const char* to_string(const PixelFormat pf);
+
+uint8_t get_bits_per_pixel(const PixelFormat pf);
+
+
+// Details about the rendered image.
+// E.g. for the 320x200 256-colour 13h VGA mode with double-scanning
+// and pixel-doubling enabled:
+//   - width = 320 (will be pixel-doubled post-render via double_width)
+//   - height = 400 (2*200 lines because we're rendering scan-doubled)
+//   - pixel_aspect_ratio = 5/6 (1:1.2) (because the PAR is meant for the
+//     final image, post the optional width & height doubling)
+//   - double_width = true (pixel-doubling)
+//   - double_height = false (we're rendering scan-doubled)
+struct RenderParams {
```

Comment:
No can't do because of circular reference problems, plus it gets propagated and stored in other parts.

It is a public structure that stores information about the rendered image, `render.cpp` "gives out" this info to other parts of the system.

But on a technical level, it needs to go into some higher level include file to break the circular reference problem. C/C++ includes and preprocessor magic... so 70s 🥶

Reply:
(`vga.h` and `render.h` both include each other; so common stuff needs to go to a third file)

Reply:
> I would say move it back or rename it.

Actually, renaming it to `ImageParams` is the best solution, you're right. It's generic enough and it's used for slightly different purposes. Done.